### PR TITLE
Divide swap into individual sell_base, sell_quote, and sell_base_to_base

### PR DESCRIPTION
**Pull request by toprince**
PR for audit issue 73

https://github.com/sherlock-audit/2024-08-woofi-solana-deployment-judging/issues/73

Divide swap into individual sell_base, sell_quote, and sell_base_to_base

### CODE REVIEW

File: programs/woofi/src/instructions/swap.rs [hunk 1]
```diff
@@ -96,6 +96,211 @@ pub struct Swap<'info> {
...
+    if ctx.accounts.woopool_from.key() == ctx.accounts.woopool_quote.key() {
+        sell_quote(ctx, from_amount, min_to_amount)
+    } else if ctx.accounts.woopool_to.key() == ctx.accounts.woopool_quote.key() {
+        sell_base(ctx, from_amount, min_to_amount)
+    } else {
+        swap_base_to_base(ctx, from_amount, min_to_amount)
+    }
+}
+
+pub fn sell_quote(ctx: Context<Swap>, from_amount: u128, min_to_amount: u128) -> Result<()> {
```
Comment (fb-alexcq) [inline programs/woofi/src/instructions/swap.rs:108]: better to be private, to align with Solidity wooPP ?

File: programs/woofi/src/instructions/swap.rs [hunk 2]
```diff
@@ -96,6 +96,211 @@ pub struct Swap<'info> {
...
+    Ok(())
+}
+
+pub fn swap_base_to_base(ctx: Context<Swap>, from_amount: u128, min_to_amount: u128) -> Result<()> {
+    require!(
+        ctx.accounts.woopool_from.token_mint != ctx.accounts.woopool_quote.token_mint
+        && ctx.accounts.woopool_to.token_mint != ctx.accounts.woopool_quote.token_mint,
+        ErrorCode::SwapPoolInvalid
+    );
+
```
Comment (fb-alexcq) [inline programs/woofi/src/instructions/swap.rs:303]: Better also add another require:

   woopool_from.token_mint != woopool_to.token_mint

File: tests/1_woofi_swap.ts [hunk 3]
```diff
@@ -451,6 +451,17 @@ describe("woofi_swap", () => {
...
+      try {
+        toPoolDataBefore = await program.account.wooPool.fetch(toPoolParams.woopool);
+      } catch (e) {
+          console.log(e);
+          console.log("fetch failed");
+          return;
+      }
+      console.log("toPool reserve before swap:" + toPoolDataBefore.reserve);
+      console.log("toPool unclaimed fee before swap:" + toPoolDataBefore.unclaimedFee);
+
```
Comment (fb-alexcq) [inline tests/1_woofi_swap.ts:464]: Why not add some `assertEquals` for reserve checks?

**Review by fb-alexcq — CHANGES_REQUESTED** [review]
some inline comments.

File: programs/woofi/src/instructions/swap.rs (hunk 2 above)
Comment (toprince) [reply programs/woofi/src/instructions/swap.rs:303]: will add

File: tests/1_woofi_swap.ts (hunk 3 above)
Comment (toprince) [reply tests/1_woofi_swap.ts:464]: Ln 611:         assert.equal(fromPoolDataAfter.reserve.toNumber(), (fromPoolDataBefore.reserve.add(new BN(fromAmount)).sub(new BN(swapFee))).toNumber())

checked.

File: programs/woofi/src/instructions/swap.rs (hunk 1 above)
Comment (toprince) [reply programs/woofi/src/instructions/swap.rs]: make 3 fn to private now.

**Review by fb-alexcq — APPROVED** [review]

**Pull request merged**